File: sample-scripts/api-basics/Generate_Token.mjs
/**
 * 
 * MIT License
 * 
 * Copyright (c) 2022 Sonetel AB
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 */
import http from 'https';
import FormData from 'form-data';

/**
 * Sample script to generate API access token for your Sonetel account.
 * 
 * Import the module into your script and call it as shown below.
 * 
 * ```javascript
 * import {createAccessToken} from 'Generate_Token.mjs';
 * const token = await createAccessToken("user@example.com", "your_password");
 * ```
 * 
 * Read more at https://docs.sonetel.com
 * 
 * @param {string} username The email address used to sign in at sonetel.com
 * @param {string} password Login password for the Sonetel account.
 */
export const createAccessToken = (username, password) => {
  return new Promise(resolve => {
    const host = 'api.sonetel.com';
    const authUri = '/SonetelAuth/beta/oauth/token';

    if (!username){
        throw Error('Username is reqired to generate access token.');
    }
    if (!password){
        throw Error('Password is reqired to generate access token.');
    }

    const requestBody = new FormData();
    requestBody.append('grant_type', 'password');
    requestBody.append('username', username);
    requestBody.append('password', password);
    requestBody.append('refresh', 'yes');

    const options = {
        hostname: host,
        path: authUri,
        auth: 'sonetel-api:sonetel-api',
        method: 'post',
        headers: requestBody.getHeaders()
      }
      const req = http.request(options, (res) => {
        const { statusCode } = res;
        if(statusCode >= 500){
          throw Error (`Request failed with server error. Status code ${statusCode}`)
        }

        res.setEncoding('utf-8');

        let rawData = '';
        res.on('data', (chunk) => rawData += chunk);
        res.on('end', async () => {
          
          try {

            let dataToReturn;
            if (res.headers['content-type'].match(/application\/json/i)) {
              dataToReturn = JSON.parse(rawData);
            } else {
              dataToReturn = rawData;
            }

            resolve(dataToReturn);

          } catch (e) {
            console.error(e.message);
          }

        });
      });
    
      req.on('error', (e) => {
        throw Error(`Request failed. ${e.message}`);
      });

      requestBody.pipe(req);
  });
}

File: sample-scripts/api-basics/Generate_Token.py
"""
Use your Sonetel username and password to generate an API token. The token can then be used as a bearer token in other API calls to buy a number, make calls, change the call forwarding settings and so on.

This example uses the Content-Type 'application/x-www-form-urlencoded' to demonstrate that your application can use either 'multipart/form-data' or 'application/x-www-form-urlencoded' when requesting for an access token.

API Documentation: https://docs.sonetel.com/docs/sonetel-documentation/YXBpOjExMzI3NDM3-authentication

Sonetel Developer Home: https://sonetel.com/en/developer/

"""
import urllib.parse
import requests
import os

# For security reasons, we recommend that you DO NOT hard code your credentials into scripts.
# Here we have assumed that your username and password are stored in your operating system's environment variables.
user = os.environ.get('sonetelUserName')
pswd = os.environ.get('sonetelPassword')

url = "https://api.sonetel.com/SonetelAuth/oauth/token"

# Prepare the request body
# If you do not want a refresh_token, remove the line 'refresh':'yes' from the request.
payload = urllib.parse.urlencode({ 
    'grant_type' : 'password',
    'password' : pswd,
    'refresh':'yes',
    'username' :  user
    })

# Request headers
headers = {
  'Accept': 'application/json, text/plain',
  'Content-Type': 'application/x-www-form-urlencoded'
}

try:
    """
    When generating the token, your Sonetel email address and password are sent in the request body. However, you also need to supply HTTP BASIC authentication with the request using sonetel-api as the username as well as the password.
    """
    response = requests.request(
        "POST", 
        url,
        auth=('sonetel-api', 'sonetel-api'), 
        data=payload, 
        headers=headers
        )
    
    # Raise an error if the API request failed.
    response.raise_for_status()
    # Print the result if the API request was successful.
    token = response.json()
    
    # Use the token, to get basic information about the Sonetel account.
    accountInfo = requests.get(
        url = 'https://public-api.sonetel.com/account/',
        headers = { "Authorization" : f'Bearer {token["access_token"]}' }
        )
    accountInfo.raise_for_status()
    print(accountInfo.json())

except requests.exceptions.RequestException as e:
    raise SystemExit(e)
